core/src/test/java/hudson/util/AtomicFileWriterTest.java

```diff
@@ -0,0 +1,108 @@
+package hudson.util;
+
+import org.hamcrest.core.StringContains;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import static org.hamcrest.core.StringContains.*;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class AtomicFileWriterTest {
+    @Rule
+    public TemporaryFolder tmp = new TemporaryFolder();
+    File af;
+    AtomicFileWriter afw;
+    String expectedContent = "hello world";
+
+    @Before
+    public void setUp() throws IOException {
+        af = tmp.newFile();
+        afw = new AtomicFileWriter(af.toPath(), Charset.defaultCharset());
+    }
+
+    @Test
+    public void symlinkToDirectory() throws Exception {
+        final File folder = tmp.newFolder();
+        final File containingSymlink = tmp.newFolder();
+        final Path zeSymlink = Files.createSymbolicLink(Paths.get(containingSymlink.getAbsolutePath(), "ze_symlink"),
+                                                         folder.toPath());
+
+
+        final Path childFileInSymlinkToDir = Paths.get(zeSymlink.toString(), "childFileInSymlinkToDir");
+
+        new AtomicFileWriter(childFileInSymlinkToDir, Charset.forName("UTF-8"));
+    }
+
+    @Test
+    public void createFile() throws Exception {
+        // Verify the file we created exists
+        assertTrue(Files.exists(afw.getTemporaryPath()));
+    }
+
+    @Test
+    public void writeToAtomicFile() throws Exception {
+        // Given
+        afw.write(expectedContent, 0, expectedContent.length());
+
+        // When
+        afw.flush();
+
+        // Then
+        assertEquals("File writer did not properly flush to temporary file",
+                expectedContent.length(), Files.size(afw.getTemporaryPath()));
+    }
+
+    @Test
+    public void commitToFile() throws Exception {
```

Comment:
What about trying to reproduce the non-Atomic write issue in a test?

I think you could try to generate an `IndexOutOfBoundsException` using `write(char cbuf[], int off, int len)`, then catch it and check the original file is intact.